webapp/src/main/java/com/box/l10n/mojito/service/thirdparty/ThirdPartyTMSSmartling.java

```diff
@@ -91,11 +137,106 @@ public void createImageToTextUnitMappings(String projectId, List<ThirdPartyImage
         smartlingClient.createBindings(bindings, projectId);
     }
 
-    Pattern getFilePattern(String repositoryName) {
-        return Pattern.compile(repositoryName + "/(\\d+)_(singular|plural)_source.xml");
+    @Override
+    public void push(Repository repository,
+                     String projectId,
+                     String pluralSeparator,
+                     String skipTextUnitsWithPattern,
+                     String skipAssetsWithPathPattern,
+                     List<String> optionList) {
+
+        lastPushResult = new ArrayList<>();
+        List<String> localeTags = ImmutableList.of("en");
+        SmartlingOptions options = SmartlingOptions.parseList(optionList);
+        AndroidStringDocumentMapper mapper = new AndroidStringDocumentMapper(pluralSeparator, null);
+
+        AtomicInteger batchCount = new AtomicInteger(0);
+        partitionSingulars(repository.getId(), localeTags, skipTextUnitsWithPattern, skipAssetsWithPathPattern)
+                .forEach(res -> processPushBatch(res, batchCount, mapper, repository, projectId, options,
+                        lastPushResult::add, "singular"));
+
+        batchCount.set(0);
+        partitionPlurals(repository.getId(), localeTags, skipTextUnitsWithPattern, skipAssetsWithPathPattern)
+                .forEach(res -> processPushBatch(res, batchCount, mapper, repository, projectId, options,
+                        lastPushResult::add, "plural"));
     }
 
-    Boolean isPluralFile(String fileUri) {
-        return fileUri.endsWith("plural_source.xml");
+    void processPushBatch(List<TextUnitDTO> result,
+                          AtomicInteger batchCounter,
+                          AndroidStringDocumentMapper mapper,
+                          Repository repository,
+                          String projectId,
+                          SmartlingOptions options,
+                          Consumer<SmartlingFile> consumer,
+                          String filePrefix) {
+
+        int batchNumber = batchCounter.getAndIncrement();
+        logger.debug("Convert text units to AndroidString for asset number: {}", batchNumber);
+        SmartlingFile file = new SmartlingFile();
+        file.setFileName(SmartlingUtils.getOutputSourceFile(batchNumber, repository.getName(), filePrefix));
+
+        try {
+
+            logger.debug("Save source file to: {}", file.getFileName());
+            AndroidStringDocumentWriter writer = new AndroidStringDocumentWriter(mapper.readFromSourceTextUnits(result));
+            file.setFileContent(writer.toText());
+
+        } catch (ParserConfigurationException | TransformerException e) {
+            logger.error("An error ocurred when processing a push batch", e);
```

Comment:
re-throw an exception here?

Reply:
btw what are those exeception, just standard XML issues? looks like we won't ever do much about it... should it moved to unchecked exception in the AndroidStringDocumentWriter